### PR TITLE
cross-dataset: coordinate noise augmentation for physics-invariant representations

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Coordinate noise augmentation — adding small Gaussian noise to input 3D node positions during training — forces the model to learn representations that are robust to geometric perturbations. This is fundamentally different from point dropout (nami #2970, which removes entire nodes) and acts instead in the continuous coordinate space.

In CFD, the underlying physics is continuous and smooth. A model that over-fits to exact mesh coordinates learns spuriously sharp features tied to mesh topology rather than fluid physics. Coordinate noise acts as a geometric regularizer: the model must predict the same pressure/velocity field even when node positions shift by epsilon, which encourages it to learn physics-invariant feature extraction.

This technique is standard in 3D point cloud literature (PointNet, DGCNN) but has not been applied in our CFD surrogate pipeline. The hypothesis is that sigma=0.005–0.01 (relative to mesh coordinate scale) will improve generalization without distorting the geometry enough to change the physics.

## Instructions

Run all 4 datasets. Add coordinate noise augmentation in `train.py`: during training, after loading a batch, add `torch.randn_like(coords) * sigma` to the node coordinate tensor before passing to the model. Do NOT apply noise during validation/evaluation. Try sigma=0.01 as the primary run. Use `--wandb_group coord-noise-cross-dataset`.

Specifically:
- Locate where node coordinates (positions/mesh points) are extracted from the batch in the training loop
- Add: `if self.training: coords = coords + torch.randn_like(coords) * 0.01`
- Do not modify validation code paths
- Set `--wandb_group coord-noise-cross-dataset`

**Dataset 1 — TandemFoil:**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --wandb_group coord-noise-cross-dataset
```

**Dataset 2 — TandemFoil Paper:**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --wandb_group coord-noise-cross-dataset
```

**Dataset 3 — AirfRANS:**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 10 --grad-clip 1.0 \
  --ema-decay 0.999 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --wandb_group coord-noise-cross-dataset
```

**Dataset 4 — DrivAerML:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group coord-noise-cross-dataset
```

Note: DrivAerML coordinates are in meters (car geometry ~4m long), so sigma=0.01m is about 1cm noise — physically meaningful. If DrivAerML diverges, try sigma=0.005.

If results look promising, suggested follow-ups: sigma=0.005, sigma=0.02, or annealing the noise from sigma_max to 0 over training.

## Baseline

Current best metrics:
- TandemFoil `val_primary/surface_pressure_mae`: **26.06** at ep300 (PR #2887, gohan)
- TandemFoil Paper `val_primary/field_mse`: TBD (racing jin #2947, guts #2948, vash #2949)
- AirfRANS `val_primary/surface_mse`: **0.000627** at ep661 (PR #2902, stark)
- DrivAerML `val_primary/surface_rel_l2_pct`: **3.997%** at ep467 (PR #2898, piccolo)

Report best-checkpoint metric per dataset. Note whether the noise helped most on TandemFoil (smallest dataset, most regularization benefit expected) vs AirfRANS/DrivAerML (larger, may be less regularization-constrained).

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shoya:

## Results

### Coordinate Noise Augmentation (sigma=0.01) — Cross-Dataset

**Summary: Coordinate noise augmentation with sigma=0.01 degrades performance significantly across all 3 tested datasets. The hypothesis did not hold.**

| Dataset | Metric | Baseline | Coord Noise (best) | Best Epoch | Delta | Status |
|---------|--------|----------|-------------------|------------|-------|--------|
| TandemFoil | `val_primary/surface_pressure_mae` | **26.06** (ep300, PR #2887) | **33.22** | ep40 | +27.5% worse | Plateaued ~33-34 |
| AirfRANS | `val_primary/surface_mse` | **0.000627** (ep661, PR #2902) | **0.001526** | ep124 | 2.43x worse | Plateaued |
| DrivAerML | `val_primary/surface_rel_l2_pct` | **3.997%** (ep467, PR #2898) | **11.45%** | ep82 | 2.86x worse | Still improving |

**Note:** TandemFoil Paper was skipped — no data files available (same finding as PR #2935, #2960).

**Note:** Experiments are still running (TF ep40, AF ep133, DR ep83, ~2.5h into 6h timeout). DrivAerML is still actively improving and may reach ~8-10% by timeout, but will not approach the 3.997% baseline. TF and AF have plateaued.

### W&B Run IDs
- TandemFoil: [`xu6a1yte`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/xu6a1yte)
- AirfRANS: [`aro88hap`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/aro88hap)
- DrivAerML: [`04wn9v3s`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/04wn9v3s)

W&B group: `coord-noise-cross-dataset`

### Peak Memory Usage
- TandemFoil (GPU 0): ~55.4 GB
- AirfRANS (GPU 1): ~25.0 GB
- DrivAerML (GPUs 2-7): ~6.0 GB per GPU

### Commands Used

**TandemFoil:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --ema-decay 0.999 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --coord-noise-sigma 0.01 \
  --epochs 999 --wandb-group coord-noise-cross-dataset \
  --wandb-name shoya/tandemfoil-coord-noise-0.01 --agent shoya
```

**AirfRANS:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 10 --grad-clip 1.0 \
  --ema-decay 0.999 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --coord-noise-sigma 0.01 \
  --epochs 999 --wandb-group coord-noise-cross-dataset \
  --wandb-name shoya/airfrans-coord-noise-0.01 --agent shoya
```

**DrivAerML:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2,3,4,5,6,7 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --coord-noise-sigma 0.01 \
  --wandb-group coord-noise-cross-dataset \
  --wandb-name shoya/drivaerml-coord-noise-0.01 --agent shoya
```

### What happened

Coordinate noise augmentation (sigma=0.01) **consistently degrades** performance across all three datasets. The degradation is substantial — ranging from 27.5% worse (TandemFoil) to 2.86x worse (DrivAerML).

**Why it likely failed:**

1. **sigma=0.01 is too large relative to mesh feature scales.** In CFD meshes, the spatial resolution near boundary layers and trailing edges encodes critical physics. Even small perturbations (0.01 in normalized coordinates) can shift points across boundary-layer regions or alter the effective geometry enough to corrupt pressure/velocity gradients that the model needs to learn.

2. **The Transolver's positional encoding is sensitive to coordinate noise.** The continuous sincos embedding (`ContinuousSincosEmbed`) maps coordinates to high-frequency Fourier features. Adding noise to coordinates before this embedding creates noise that is amplified in the high-frequency components, making it much harder for the slice-attention mechanism to form coherent physical groupings.

3. **CFD is not point cloud classification.** In PointNet/DGCNN literature, coordinate noise helps because those tasks (classification, segmentation) are naturally invariant to small perturbations. CFD regression tasks are NOT — pressure fields have sharp gradients tied to exact geometry, and perturbing coordinates fundamentally changes the problem the model is being asked to solve.

4. **DrivAerML is worst affected** (2.86x at ep82, though still improving): its 3D geometry with sigma=0.01m (≈1cm) noise is physically significant on car aerodynamics features like mirrors, A-pillars, and wheel arches.

### Suggested follow-ups

1. **Much smaller sigma (0.001 or 0.0005):** The current sigma=0.01 may simply be too aggressive. A sigma 10-20x smaller might regularize without corrupting the physics signal.

2. **Noise annealing:** Start with sigma=0.01 and anneal to 0 over training — early noise for regularization, clean training for final convergence.

3. **Apply noise only to volume points, not surface points:** Surface points carry the critical boundary condition information. Noising only the volume (far-field) points might provide regularization without corrupting the physics signal at the boundary.

4. **Feature-space noise instead of coordinate noise:** Add noise to the hidden representations after the positional embedding rather than to raw coordinates. This avoids the amplification through high-frequency Fourier features.
